src/util/path.ts

```diff
@@ -2,26 +2,133 @@ import * as vscode from 'vscode';
+    isWindows = currentUri.scheme === 'file' && currentUri.fsPath[0] !== '/';
+  }
+
+  const p = isWindows ? path.win32 : path.posix;
+  if (isWindows) {
+    // normalize / to \ on windows
+    partialPath = partialPath.replace(/\//g, '\\');
+  }
+  const updatedPartialPath = partialPath;
+
+  if (currentUri.scheme === 'file' || (currentUri.scheme === 'untitled' && !isRemote)) {
+    // We can untildify when the scheme is 'file' or 'untitled' on local fs because
+    // because we only support opening files mounted locally.
+    partialPath = untildify(partialPath);
   }
 
-  return basePath + partialPath;
+  let [dirName, baseName] = separatePath(partialPath, p.sep);
+  let fullDirPath: string;
+  if (p.isAbsolute(dirName)) {
+    fullDirPath = dirName;
+  } else {
+    fullDirPath = p.join(
+      // On Windows machine:
+      // fsPath returns Windows drive path (C:\xxx\) or UNC path (\\server\xxx)
+      // fsPath returns path with \ as separator even if 'vscode-remote' is connect to a linux box
+      //
+      // path will return /home/user for example even 'vscode-remote' is used on windows
+      // as we relied of our isWindows detection
+      separatePath(isWindows ? currentUri.fsPath : currentUri.path, p.sep)[0],
+      dirName
+    );
+  }
+
+  const fullPath = p.join(fullDirPath, baseName);
+  return {
+    fullPath,
+    fullDirPath,
+    dirName,
+    baseName,
+    partialPath: updatedPartialPath,
+    path: p,
+  };
+}
+
+export function getUriPath(currentUri: vscode.Uri, absolutePath: string, sep: string) {
+  const isWindows = sep === path.win32.sep;
+  if (isWindows && !/^(\\\\.+\\)|([a-zA-Z]:\\)/.test(absolutePath)) {
+    // if it is windows and but don't have either
+    // UNC path or the windows drive
+    return null;
+  }
+  if (!isWindows && absolutePath[0] !== sep) {
+    // if it is not windows, but the absolute path doesn't begin with /
+    return null;
+  }
+
+  const isRemote = !!vscode.env.remoteName;
+  const isLocalUntitled = !isRemote && currentUri.scheme === 'untitled';
+  return isWindows
+    ? // create new local Uri when it's on windows (doesn't support remote)
+      // Use file will also works for UNC paths like //server1/folder
+      vscode.Uri.file(absolutePath)
+    : currentUri.with({
+        // search local file with it's untitled and not remote
+        scheme: isLocalUntitled ? 'file' : currentUri.scheme,
+        path: absolutePath,
+      });
+}
+
+export async function readDirectory(
+  currentUri: vscode.Uri,
+  absolutePath: string,
+  sep: string,
+  addCurrentAndUp: boolean
+) {
+  try {
+    const directoryUri = getUriPath(currentUri, absolutePath, sep);
+    if (directoryUri === null) {
+      return [];
+    }
+    const directoryResult = await vscode.workspace.fs.readDirectory(directoryUri);
+    return directoryResult
+      .map(d => ({ path: d[0] + (d[1] === vscode.FileType.Directory ? sep : ''), type: d[1] }))
```

Comment:
can't figure out what's going on here...

Reply:
f[0] is the item name in a directory, and d[1] is the type of the item.
Basically, I am appending separator at the end as the indicator of directory item.
I refactored this method a bit so it only returns an array of strings.